### PR TITLE
Reducing memory pressure in RawCaptureSocket

### PR DESCRIPTION
**Pull request by sandtechnology**
Cut off useless buffer when is less than half of buffer size for saving memory

### CODE REVIEW

File: Machina/Sockets/RawCaptureSocket.cs [hunk 1]
```diff
@@ -139,7 +139,19 @@ private void OnReceive(IAsyncResult ar)
                     _ = _socket.BeginReceive(_currentBuffer, 0, _currentBuffer.Length, SocketFlags.None, new AsyncCallback(OnReceive), null);
 
                     if (received > 0)
-                        _pendingBuffers.Enqueue(new Tuple<byte[], int>(buffer, received));
+                    {
+                        //Cut off useless buffer when is less than half of buffer size for saving memory
+                        if (BUFFER_SIZE / 2 > received)
```
Comment (ravahn) [inline Machina/Sockets/RawCaptureSocket.cs:144]: Can you write more about what the goal of this PR is?

The code you added here will be executed on every network receive, and will allocate a new byte array.   This increases memory allocations.  if the goal is to prevent extra bytes at the end of the buffer, a better way would be to not reallocate _currentBuffer in the lines above, and just copy the received bytes (assuming received>0) to a new array and enque that before calling BeginReceive().  But, I didn't do that because the code was previously pulling buffers from a fixed-length buffer pool, now removed due to other concerns.  And, that will introduce slightly more time where the socket kernel buffer will receive data before it is polled again.

Can you describe where you observed issues with 64kb buffers?

Comment (sandtechnology) [reply Machina/Sockets/RawCaptureSocket.cs:144]: Thanks for looking this PR and your suggestion! My goal is prevent extra unused bytes in buffer, in there it cost more cpu time for lesser space occupied in `_pendingbuffer`, so when there are so many small packet (e.g 50B packets), the memory it used won't growing repaidly. (The origin purpose is, in FFXIV process, ghub will create an connection to its agent for implementing its feature, and its packets will also captured by machina, and look like it using LOTS of packet)

Also I will optimize it according to your suggestion, and I will update the goal in description.

Comment (sandtechnology) [reply Machina/Sockets/RawCaptureSocket.cs:141]: Done, now is ready for reviewing.

**Comment by ravahn** [comment]
You make a good case.  I'm going to take a slightly different approach to solve this, to eliminate some duplicate code and a possible extra allocation for large buffers:

Here is a version of this which is a little more concise:
 ```
int received = _socket.EndReceive(ar);
if (received > 0)
{
  byte[] buffer = new byte[received];
  Array.Copy(_currentBuffer, 0, buffer, 0, received);

  _pendingBuffers.Enqueue(new Tuple<byte[], int>(buffer, received));
}
_ = _socket.BeginReceive(_currentBuffer, 0, _currentBuffer.Length, SocketFlags.None, new AsyncCallback(OnReceive), null);

```
I'll also try to simplify _pendingBuffers to eliminate the length parameter.

**Comment by ravahn** [comment]
Take a look at the code I just pushed to master branch and let me know if it improves the memory allocations for you.  Thanks!

**Comment by sandtechnology** [comment]
Looks like is great and should do the same purpose since it using mostly same logic, and it should more efficient since it remove redundant `Tuple`.

The reason why I put receive code before `Enqueue` to queue is because `Enqueue` method may run a while for adding element, it using CAS(Compare And Swap) algorithm:
```
		public void Enqueue(T item)
		{
			SpinWait spinWait = default(SpinWait);
			for (;;)
			{
				ConcurrentQueue<T>.Segment tail = this.m_tail;
				if (tail.TryAppend(item))
				{
					break;
				}
				spinWait.SpinOnce();
			}
		}
```

it's what your origin code does, so I also put the receive code before `Enqueue` invoke to follow this logic, but to make the code more simpler, this optimization could be also removed since it may not run into such corner case (the thread using it should be not that much in most cases), and the packet received should be in order.

Anyway, I will test it for a while and post a reply after that, thanks for your great work!
